source/adapters/level_zero/v2/event.cpp

```diff
@@ -7,28 +7,49 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
+
+#include <ze_api.h>
+
 #include "event.hpp"
+#include "event_pool.hpp"
 #include "event_provider.hpp"
-#include "ze_api.h"
 
-namespace v2 {
-void ur_event::attachZeHandle(event_allocation event) {
-  type = event.type;
-  zeEvent = std::move(event.borrow);
-}
+ur_event_handle_t_::ur_event_handle_t_(v2::event_allocation eventAllocation,
+                                       v2::event_pool *pool)
+    : type(eventAllocation.type), zeEvent(std::move(eventAllocation.borrow)),
+      pool(pool) {}
 
-event_borrowed ur_event::detachZeHandle() {
+void ur_event_handle_t_::reset() {
   // consider make an abstraction for regular/counter based
   // events if there's more of this type of conditions
-  if (type == event_type::EVENT_REGULAR) {
+  if (type == v2::event_type::EVENT_REGULAR) {
     zeEventHostReset(zeEvent.get());
   }
-  auto e = std::move(zeEvent);
-  zeEvent = nullptr;
+}
+
+ze_event_handle_t ur_event_handle_t_::getZeEvent() const {
+  return zeEvent.get();
+}
 
-  return e;
+ur_result_t ur_event_handle_t_::retain() {
+  RefCount.increment();
+  return UR_RESULT_SUCCESS;
 }
 
-ze_event_handle_t ur_event::getZeEvent() { return zeEvent.get(); }
+ur_result_t ur_event_handle_t_::release() {
+  if (!RefCount.decrementAndTest())
+    return UR_RESULT_SUCCESS;
 
-} // namespace v2
+  pool->free(this);
+  RefCount.increment();
+
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urEventRetain(ur_event_handle_t hEvent) {
+  return hEvent->retain();
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urEventRelease(ur_event_handle_t hEvent) {
+  return hEvent->release();
```

Comment:
This is not MT-safe. The hEvent needs to store a queue, not the pool, and then grab the queue lock to access its pool. In debug we could add a check to make sure that the queue's pool is exactly the same as the event's pool.

I know we can't implement that right now, so let's just leave this as TODO (no need to push another revision of PR :P I don't want to waste time on CI).

Reply:
Right, event_pool is not thread safe. I'll make a note